packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -0,0 +1,183 @@
+  | TSESTree.DoWhileStatement
+  | TSESTree.ForStatement
+  | TSESTree.IfStatement
+  | TSESTree.WhileStatement;
+
+export type Options = [
+  {
+    ignoreRhs?: boolean;
+  },
+];
+
+export type MessageId =
+  | 'alwaysTruthy'
+  | 'alwaysFalsy'
+  | 'literalBooleanExpression'
+  | 'never';
+export default createRule<Options, MessageId>({
+  name: 'no-unnecessary-conditionals',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Prevents conditionals where the type is always truthy or always falsy',
+      category: 'Best Practices',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreRhs: {
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+    messages: {
+      alwaysTruthy: 'Unnecessary conditional, value is always truthy.',
+      alwaysFalsy: 'Unnecessary conditional, value is always falsy.',
+      literalBooleanExpression:
+        'Unnecessary conditional, both sides of the expression are literal values',
+      never: 'Unnecessary conditional, value is `never`',
+    },
+  },
+  defaultOptions: [
+    {
+      ignoreRhs: false,
+    },
+  ],
+  create(context, [{ ignoreRhs }]) {
+    const service = getParserServices(context);
+    const checker = service.program.getTypeChecker();
+
+    function getNodeType(node: TSESTree.Node): ts.Type {
+      const tsNode = service.esTreeNodeToTSNodeMap.get(node);
+      return getConstrainedTypeAtLocation(checker, tsNode);
+    }
+
+    /**
+     * Checks if a conditional node is necessary:
+     * if the type of the node is always true or always false, it's not necessary.
+     */
+    function checkNode(node: TSESTree.Node): void {
+      const type = getNodeType(node);
+
+      // Conditional is always necessary if it involves `any` or `unknown`
+      if (isTypeFlagSet(type, TypeFlags.Any | TypeFlags.Unknown)) {
+        return;
+      }
+      if (isTypeFlagSet(type, TypeFlags.Never)) {
+        context.report({ node, messageId: 'never' });
+      } else if (!isPossiblyTruthy(type)) {
+        context.report({ node, messageId: 'alwaysFalsy' });
+      } else if (!isPossiblyFalsy(type)) {
+        context.report({ node, messageId: 'alwaysTruthy' });
+      }
+    }
+
+    /**
+     * Checks that a binary expression is necessarily conditoinal, reports otherwise.
+     * If both sides of the binary expression are literal values, it's not a necessary condition.
+     *
+     * NOTE: It's also unnecessary if the types that don't overlap at all
+     *    but that case is handled by the Typescript compiler itself.
+     */
+    const BOOL_OPERATORS = ['<', '>', '<=', '>=', '==', '===', '!=', '!=='];
+    function checkIfBinaryExpressionIsNecessaryConditional(
+      node: TSESTree.BinaryExpression,
+    ): void {
+      if (
+        BOOL_OPERATORS.includes(node.operator) &&
```

Comment:
For matching set change
```suggestion
        BOOL_OPERATORS.has(node.operator) &&
```